kernel/comps/keyboard/src/i8042_chip/mouse.rs

```diff
@@ -0,0 +1,474 @@
+    let Some(data) = controller.lock().receive_data() else {
+        log::warn!("PS/2 mouse has no input data");
+        return;
+    };
+
+    let mut packet_state = PACKET_STATE.lock();
+    if let Some(events) = packet_state.process_byte(data) {
+        if let Some(registered_device) = REGISTERED_DEVICE.get() {
+            registered_device.submit_events(&events);
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum MouseType {
+    Standard,
+    IntelliMouse,
+}
+
+impl MouseType {
+    const PACKET_LEN_STANDARD: usize = 3;
+    const PACKET_LEN_INTELLIMOUSE: usize = 4;
+    const PACKET_LEN_MAX: usize = 4;
+
+    fn from_device_id(device_id: u8) -> Option<Self> {
+        const DEVICE_ID_STANDARD_MOUSE: u8 = 0x00;
+        const DEVICE_ID_INTELLIMOUSE: u8 = 0x03;
+        const DEVICE_ID_INTELLIMOUSE_EXPLORER: u8 = 0x04;
+
+        match device_id {
+            DEVICE_ID_STANDARD_MOUSE => Some(MouseType::Standard),
+            DEVICE_ID_INTELLIMOUSE | DEVICE_ID_INTELLIMOUSE_EXPLORER => {
+                Some(MouseType::IntelliMouse)
+            }
+            _ => None,
+        }
+    }
+
+    fn packet_len(&self) -> usize {
+        match self {
+            MouseType::Standard => Self::PACKET_LEN_STANDARD,
+            MouseType::IntelliMouse => Self::PACKET_LEN_INTELLIMOUSE,
+        }
+    }
+}
+
+/// Mouse packet state machine for handling 3-byte or 4-byte PS/2 mouse packets.
+struct PacketState {
+    packet: [u8; MouseType::PACKET_LEN_MAX],
+    byte_index: usize,
+    mouse_type: MouseType,
+    prev_left: bool,
+    prev_right: bool,
+    prev_middle: bool,
+}
+
+impl PacketState {
+    const fn new() -> Self {
+        Self {
+            packet: [0; MouseType::PACKET_LEN_MAX],
+            byte_index: 0,
+            mouse_type: MouseType::Standard,
+            prev_left: false,
+            prev_right: false,
+            prev_middle: false,
+        }
+    }
+
+    fn set_mouse_type(&mut self, mouse_type: MouseType) {
+        self.mouse_type = mouse_type;
+        self.reset();
+    }
+
+    /// Processes a mouse data byte and returns input events if a complete packet is ready.
+    fn process_byte(&mut self, data: u8) -> Option<Vec<InputEvent>> {
+        const PACKET_SYNC_MASK: u8 = 0x08;
+
+        // The first byte must have bit 3 set (always 1 in PS/2 mouse packets).
+        if self.byte_index == 0 && (data & PACKET_SYNC_MASK) == 0 {
+            return None;
+        }
+
+        // Collect the byte into the packet.
+        self.packet[self.byte_index] = data;
+        self.byte_index += 1;
+        if self.byte_index < self.mouse_type.packet_len() {
+            return None;
+        }
+
+        // A complete packet has been received.
+        let events = self.parse_packet();
+        self.reset();
+        Some(events)
+    }
+
+    /// Parses a complete PS/2 mouse packet and returns input events.
+    /// Reference: <https://wiki.osdev.org/Mouse_Input#Format_of_First_3_Packet_Bytes>.
```

Comment:
Please add a blank line before `Reference: `. The first sentence of the doc comments should be a summary sentence. See https://rust-lang.github.io/rfcs/0505-api-comment-conventions.html#formatting.
```rust
    /// Parses a complete PS/2 mouse packet and returns input events.
    ///
    /// Reference: <https://wiki.osdev.org/Mouse_Input#Format_of_First_3_Packet_Bytes>.
```